src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -1343,7 +1343,7 @@ private void StopWatchingForExit()
 
         public override string ToString()
         {
-            if (Associated)
+            if (Associated && !HasExited)
```

Comment:
What happens if the process exits after checking HasExited and before accessing ProcessName?

Reply:
Good point. Catch?

Reply:
That's a valid question! Is it worth just wrapping this in a `try/catch` then, or is that unlikely enough to not be a concern?

Reply:
Where is the problematic exception coming from? i.e. what's the stack of the exception that's thrown if the process has exited?

Reply:
@stephentoub it's coming from the call to `ProcessName`  which then calls `EnsureState` ( https://github.com/dotnet/runtime/blob/098fdc5fe634067283369bbb51e7c439431e4105/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L512) - the state flag it uses means it'll call `EnsureState` again, this time checking that it has a non-exited process ID (https://github.com/dotnet/runtime/blob/098fdc5fe634067283369bbb51e7c439431e4105/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L951). That's where the exception gets thrown.

Reply:
> That's where the exception gets thrown.

Hmm, are you sure?  It's passing in State.HaveProcessInfo, which doesn't include the bits for HaveNonExitedId, so I would not expect it to be making a call to EnsureState(State.HaveNonExitedId).  Rather, I'd expect the exception is coming from here:
https://github.com/dotnet/runtime/blob/21121c35e95aacc0083d794213aef0f2af7b9296/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L953-L957
with GetProcessInfo returning null and the call then explicitly throwing as a result (the error string thrown also matches the text I see in the exception.

Assuming that's true, then I think I'd prefer to see ToString just rewritten to not use ProcessName at all and avoid the throwing altogether:
```C#
public override string ToString()
{
    string result = base.ToString();

    if (Associated)
    {
        _processInfo ??= ProcessManager.GetProcessInfo(_processId, _machineName);
        if (_processInfo is not null)
        {
            string processName = _processInfo.ProcessName;
            if (processName.Length != 0)
            {
                result = $"{result} ({processName})";
            }
        }
    }

    return result;
}
```

Reply:
What about
```c#
public override string ToString()
{
    string result = null;

    if (Associated)
    {
        _processInfo ??= ProcessManager.GetProcessInfo(_processId, _machineName);
        if (_processInfo is not null)
        {
            string processName = _processInfo.ProcessName;
            if (processName.Length != 0)
            {
                result = $"{result} ({processName})";
            }
        }
    }

    return result ?? base.ToString();
}
```

Reply:
@iSazonov I think, considering I need the `base.ToString()` in _both_ results, it's fine if I call it in the beginning.

Reply:
@stephentoub I feel a little bit "bad" setting `_processInfo` directly from a `ToString()` method, though. My thinking was the `!HasExisted` check would be good enough, since it then looks at `if ((state & State.HaveProcessInfo) != (State)0)`.  That way the `processInfo` gets set in the same place you'd expect it to. 

Happy to go either way, though. 

But yeah, if it exists immediately after that check, maybe it makes sense to do a `try/catch`.

Reply:
I've gone ahead with the changes, and rebased while I'm at it.